File: ts/src/quorum/opencode-capture.ts
/**
 * Export OpenCode sessions from isolated per-run state.
 *
 * Port of quorum/opencode_capture.py — public API is camelCase TS, logic is identical.
 */

import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

// ---------------------------------------------------------------------------
// Error
// ---------------------------------------------------------------------------

export class OpenCodeCaptureError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "OpenCodeCaptureError";
  }
}

// ---------------------------------------------------------------------------
// Environment helpers
// ---------------------------------------------------------------------------

export function opencodeEnv(home: string): Record<string, string> {
  return {
    HOME: home,
    XDG_CONFIG_HOME: path.join(home, ".config"),
    XDG_DATA_HOME: path.join(home, ".local", "share"),
    XDG_STATE_HOME: path.join(home, ".local", "state"),
    XDG_CACHE_HOME: path.join(home, ".cache"),
    TMPDIR: path.join(home, ".tmp"),
    OPENCODE_CONFIG_DIR: path.join(home, ".config", "opencode"),
  };
}

export const OPENCODE_ENV_ALLOWLIST = new Set([
  "PATH",
  "TERM",
  "COLORTERM",
  "LANG",
  "LC_ALL",
  "SSL_CERT_FILE",
  "REQUESTS_CA_BUNDLE",
  "NODE_EXTRA_CA_CERTS",
  "OPENAI_API_KEY",
  "ANTHROPIC_API_KEY",
  "OPENROUTER_API_KEY",
  "GEMINI_API_KEY",
  "GOOGLE_API_KEY",
  "AWS_PROFILE",
  "AWS_REGION",
  "AWS_DEFAULT_REGION",
  "AWS_ACCESS_KEY_ID",
  "AWS_SECRET_ACCESS_KEY",
  "AWS_SESSION_TOKEN",
]);

export const OPENCODE_CAPTURE_TIMEOUT_MS = 30_000;

export function opencodeRunEnv(home: string): Record<string, string> {
  const env: Record<string, string> = {};
  for (const [key, value] of Object.entries(process.env)) {
    if (OPENCODE_ENV_ALLOWLIST.has(key) && value !== undefined) {
      env[key] = value;
    }
  }
  if (!("PATH" in env)) env["PATH"] = "/usr/bin:/bin";
  if (!("TERM" in env)) env["TERM"] = "xterm-256color";
  if (!("LANG" in env)) env["LANG"] = "C.UTF-8";
  Object.assign(env, opencodeEnv(home));
  return env;
}

// ---------------------------------------------------------------------------
// Spawn function type
// ---------------------------------------------------------------------------

export interface SpawnResult {
  stdout: string;
  stderr: string;
  exitCode: number;
}

/**
 * Injectable spawn function type. Takes the command args, cwd, env, and
 * timeout (ms). Returns SpawnResult synchronously.
 */
export type SpawnFn = (opts: {
  args: string[];
  cwd: string;
  env: Record<string, string>;
  timeoutMs: number;
}) => SpawnResult;

// ---------------------------------------------------------------------------
// Default spawn implementation (uses Bun.spawnSync + temp file for stdout)
// ---------------------------------------------------------------------------

/**
 * Run an opencode CLI command with stdout redirected to a temp file.
 *
 * The opencode binary ends every command with a bare process.exit(), which
 * discards stdout that has not yet drained. Through a pipe, payloads >64KiB
 * arrive truncated at the pipe-buffer boundary (still exit 0) and tiny
 * replies can vanish entirely under load. A regular-file stdout drains
 * synchronously, so the payload survives. stderr stays piped (always small).
 */
export function defaultSpawn(opts: {
  args: string[];
  cwd: string;
  env: Record<string, string>;
  timeoutMs: number;
}): SpawnResult {
  const { args, cwd, env, timeoutMs } = opts;
  const tmpFile = path.join(
    os.tmpdir(),
    `opencode-stdout-${process.pid}-${Date.now()}-${Math.random().toString(36).slice(2)}.tmp`,
  );
  fs.writeFileSync(tmpFile, "");
  try {
    const stdoutFd = fs.openSync(tmpFile, "r+");
    try {
      const proc = Bun.spawnSync(args, {
        cwd,
        env,
        stdin: "ignore",
        stdout: stdoutFd,
        stderr: "pipe",
        timeout: timeoutMs,
      });
      const stdout = fs.readFileSync(tmpFile, "utf8");
      const stderr =
        proc.stderr instanceof Uint8Array
          ? new TextDecoder().decode(proc.stderr)
          : typeof proc.stderr === "string"
            ? proc.stderr
            : "";
      const exitCode = proc.exitCode ?? 0;
      return { stdout, stderr, exitCode };
    } finally {
      fs.closeSync(stdoutFd);
    }
  } finally {
    try {
      fs.unlinkSync(tmpFile);
    } catch {
      // ignore cleanup errors
    }
  }
}

// ---------------------------------------------------------------------------
// runOpencodeCommand
// ---------------------------------------------------------------------------

export function runOpencodeCommand(
  args: string[],
  opts: {
    opencodeHome: string;
    launchCwd: string;
    timeoutMs?: number;
    spawn?: SpawnFn;
  },
): SpawnResult {
  const { opencodeHome, launchCwd, timeoutMs = OPENCODE_CAPTURE_TIMEOUT_MS, spawn = defaultSpawn } =
    opts;
  const result = spawn({
    args: ["opencode", ...args],
    cwd: launchCwd,
    env: opencodeRunEnv(opencodeHome),
    timeoutMs,
  });
  return result;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

export function _realpath(value: string): string {
  // Mirror Python's os.path.realpath: resolves symlinks but does not throw
  // for non-existent paths (unlike fs.realpathSync which throws ENOENT).
  try {
    return fs.realpathSync(value);
  } catch {
    // For non-existent paths, normalize without resolving symlinks.
    return path.resolve(value);
  }
}

interface SessionRow {
  id: string;
  directory: string;
  [key: string]: unknown;
}

interface SessionDecision {
  index: number;
  id?: unknown;
  directory?: string;
  directory_realpath?: string;
  launch_cwd_realpath?: string;
  matched: boolean;
  reason?: string;
}

function sessionDecisions(
  rawSessions: unknown,
  launchCwd: string,
): [SessionDecision[], SessionRow[]] {
  if (!Array.isArray(rawSessions)) {
    throw new OpenCodeCaptureError("opencode session list returned non-list JSON");
  }
  const target = _realpath(launchCwd);
  const decisions: SessionDecision[] = [];
  const matches: SessionRow[] = [];
  for (let index = 0; index < rawSessions.length; index++) {
    const session = rawSessions[index];
    if (typeof session !== "object" || session === null || Array.isArray(session)) {
      decisions.push({ index, matched: false, reason: "non-dict row" });
      continue;
    }
    const sessionRow = session as Record<string, unknown>;
    const directory = sessionRow["directory"];
    const sessionId = sessionRow["id"];
    if (typeof directory !== "string" || typeof sessionId !== "string") {
      decisions.push({
        index,
        id: sessionId,
        matched: false,
        reason: "missing id or directory",
      });
      continue;
    }
    const directoryRealpath = _realpath(directory);
    const matched = directoryRealpath === target;
    decisions.push({
      index,
      id: sessionId,
      directory,
      directory_realpath: directoryRealpath,
      launch_cwd_realpath: target,
      matched,
    });
    if (matched) {
      matches.push(sessionRow as unknown as SessionRow);
    }
  }
  return [decisions, matches];
}

function listSessions(opts: {
  opencodeHome: string;
  launchCwd: string;
  spawn?: SpawnFn;
}): unknown[] {
  const { opencodeHome, launchCwd, spawn } = opts;
  let result: SpawnResult;
  try {
    result = runOpencodeCommand(["session", "list", "--format", "json"], {
      opencodeHome,
      launchCwd,
      spawn,
    });
  } catch (e) {
    if (isTimeoutError(e)) {
      throw new OpenCodeCaptureError(
        `opencode session list timed out after ${OPENCODE_CAPTURE_TIMEOUT_MS / 1000}s`,
      );
    }
    throw e;
  }
  if (result.exitCode !== 0) {
    throw new OpenCodeCaptureError(
      `opencode session list failed (exit ${result.exitCode}): ${result.stderr.trim().slice(0, 300)}`,
    );
  }
  let sessions: unknown;
  try {
    sessions = JSON.parse(result.stdout || "[]");
  } catch {
    throw new OpenCodeCaptureError("opencode session list returned invalid JSON");
  }
  if (!Array.isArray(sessions)) {
    throw new OpenCodeCaptureError("opencode session list returned non-list JSON");
  }
  return sessions;
}

function isTimeoutError(e: unknown): boolean {
  if (e instanceof Error) {
    return e.message.toLowerCase().includes("timeout") || e.constructor.name === "TimeoutError";
  }
  return false;
}

// ---------------------------------------------------------------------------
// snapshotOpencodeSessions
// ---------------------------------------------------------------------------

export function snapshotOpencodeSessions(opts: {
  home: string;
  launchCwd: string;
  spawn?: SpawnFn;
}): Set<string> {
  const { home, launchCwd, spawn } = opts;
  const rawSessions = listSessions({ opencodeHome: home, launchCwd, spawn });
  const [, sessions] = sessionDecisions(rawSessions, launchCwd);
  return new Set(sessions.map((s) => s["id"]));
}

// ---------------------------------------------------------------------------
// Session created timestamp helpers
// ---------------------------------------------------------------------------

function sessionCreated(session: Record<string, unknown>): number | null {
  for (const key of ["created", "time_created"]) {
    const value = session[key];
    if (typeof value === "number") return value;
  }
  return null;
}

function exportedCreated(exportedJson: Record<string, unknown>): number | null {
  const info = exportedJson["info"];
  if (typeof info !== "object" || info === null) return null;
  const time = (info as Record<string, unknown>)["time"];
  if (typeof time !== "object" || time === null) return null;
  const created = (time as Record<string, unknown>)["created"];
  return typeof created === "number" ? created : null;
}

// ---------------------------------------------------------------------------
// exportSession
// ---------------------------------------------------------------------------

function exportSession(opts: {
  sessionId: string;
  opencodeHome: string;
  launchCwd: string;
  spawn?: SpawnFn;
}): [Record<string, unknown>, string, string] {
  const { sessionId, opencodeHome, launchCwd, spawn } = opts;
  let result: SpawnResult;
  try {
    result = runOpencodeCommand(["export", sessionId], {
      opencodeHome,
      launchCwd,
      spawn,
    });
  } catch (e) {
    if (isTimeoutError(e)) {
      throw new OpenCodeCaptureError(
        `opencode export ${sessionId} timed out after ${OPENCODE_CAPTURE_TIMEOUT_MS / 1000}s`,
      );
    }
    throw e;
  }
  if (result.exitCode !== 0) {
    throw new OpenCodeCaptureError(
      `opencode export ${sessionId} failed (exit ${result.exitCode}): ${result.stderr.trim().slice(0, 300)}`,
    );
  }
  let exportedJson: Record<string, unknown>;
  try {
    exportedJson = JSON.parse(result.stdout) as Record<string, unknown>;
  } catch {
    const byteLen = new TextEncoder().encode(result.stdout).length;
    throw new OpenCodeCaptureError(
      `opencode export ${sessionId} returned invalid JSON ` +
        `(${byteLen} bytes; ` +
        `head: ${JSON.stringify(result.stdout.slice(0, 120))}; ` +
        `stderr: ${result.stderr.trim().slice(0, 300)})`,
    );
  }
  const info = exportedJson["info"];
  const exportedId =
    typeof info === "object" && info !== null ? (info as Record<string, unknown>)["id"] : undefined;
  if (exportedId !== sessionId) {
    throw new OpenCodeCaptureError(
      `opencode export ${sessionId} returned session id ${JSON.stringify(exportedId)}`,
    );
  }
  return [exportedJson, result.stdout, result.stderr];
}

// ---------------------------------------------------------------------------
// exportOpencodeSessions
// ---------------------------------------------------------------------------

interface ExportRecord {
  id: string;
  json: Record<string, unknown>;
  stdout: string;
  stderr: string;
  created: number | null;
}

export function exportOpencodeSessions(opts: {
  opencodeHome: string;
  exportDir: string;
  launchCwd: string;
  snapshot: Set<string>;
  spawn?: SpawnFn;
}): string[] {
  const { opencodeHome, exportDir, launchCwd, snapshot, spawn } = opts;
  fs.mkdirSync(exportDir, { recursive: true });

  const rawSessions = listSessions({ opencodeHome, launchCwd, spawn });
  const [decisions, sessions] = sessionDecisions(rawSessions, launchCwd);
  const newSessions = sessions.filter((s) => !snapshot.has(s["id"]));

  const exportRecords: ExportRecord[] = [];
  for (const session of newSessions) {
    const sessionId = session["id"];
    const [exportedJson, stdout, stderr] = exportSession({
      sessionId,
      opencodeHome,
      launchCwd,
      spawn,
    });
    const created =
      sessionCreated(session as Record<string, unknown>) || exportedCreated(exportedJson);
    exportRecords.push({ id: sessionId, json: exportedJson, stdout, stderr, created });
  }

  if (exportRecords.length > 1 && exportRecords.some((r) => r.created === null)) {
    throw new OpenCodeCaptureError(
      "cannot order multiple new OpenCode sessions without creation times",
    );
  }

  exportRecords.sort((a, b) => {
    const ca = a.created ?? 0;
    const cb = b.created ?? 0;
    if (ca !== cb) return ca - cb;
    return a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
  });

  const exported: string[] = [];

  const manifest: Record<string, unknown> = {
    raw_session_rows: rawSessions,
    session_decisions: decisions,
    snapshot_ids: [...snapshot].sort(),
    all_matching_ids: sessions.map((s) => s["id"]),
    matched_ids: newSessions.map((s) => s["id"]),
    skipped_existing_ids: sessions.filter((s) => snapshot.has(s["id"])).map((s) => s["id"]),
    skipped_nonmatching_ids: decisions
      .filter((d) => d.id !== undefined && !d.matched)
      .map((d) => d.id as string),
    exports: [] as unknown[],
  };

  for (const record of exportRecords) {
    const created = record.created ?? 0;
    const filename = `${created.toString().padStart(16, "0")}-${record.id}.json`;
    const outPath = path.join(exportDir, filename);
    fs.writeFileSync(outPath, record.stdout, "utf8");
    (manifest["exports"] as unknown[]).push({
      id: record.id,
      created,
      path: outPath,
      stderr: record.stderr,
    });
    exported.push(outPath);
  }

  fs.writeFileSync(
    path.join(exportDir, "opencode-session-export-manifest.json"),
    JSON.stringify(manifest, null, 2) + "\n",
    "utf8",
  );

  return exported;
}

File: quorum/opencode_capture.py
"""Export OpenCode sessions from isolated per-run state."""

from __future__ import annotations

import json
import os
import subprocess
import tempfile
from collections.abc import Mapping
from pathlib import Path
from typing import Any, cast

from quorum.runtime_env import is_managed_host


class OpenCodeCaptureError(RuntimeError):
    """Raised when OpenCode session export cannot complete."""


def opencode_env(opencode_home: Path) -> dict[str, str]:
    return {
        "HOME": str(opencode_home),
        "XDG_CONFIG_HOME": str(opencode_home / ".config"),
        "XDG_DATA_HOME": str(opencode_home / ".local" / "share"),
        "XDG_STATE_HOME": str(opencode_home / ".local" / "state"),
        "XDG_CACHE_HOME": str(opencode_home / ".cache"),
        "TMPDIR": str(opencode_home / ".tmp"),
        "OPENCODE_CONFIG_DIR": str(opencode_home / ".config" / "opencode"),
    }


OPENCODE_ENV_ALLOWLIST = {
    "PATH",
    "TERM",
    "COLORTERM",
    "LANG",
    "LC_ALL",
    "SSL_CERT_FILE",
    "REQUESTS_CA_BUNDLE",
    "NODE_EXTRA_CA_CERTS",
    "OPENAI_API_KEY",
    "ANTHROPIC_API_KEY",
    "OPENROUTER_API_KEY",
    "GEMINI_API_KEY",
    "GOOGLE_API_KEY",
    "AWS_PROFILE",
    "AWS_REGION",
    "AWS_DEFAULT_REGION",
    "AWS_ACCESS_KEY_ID",
    "AWS_SECRET_ACCESS_KEY",
    "AWS_SESSION_TOKEN",
}
OPENCODE_PROVIDER_ENV_NAMES = {
    "OPENAI_API_KEY",
    "ANTHROPIC_API_KEY",
    "OPENROUTER_API_KEY",
    "GEMINI_API_KEY",
    "GOOGLE_API_KEY",
    "AWS_PROFILE",
    "AWS_REGION",
    "AWS_DEFAULT_REGION",
    "AWS_ACCESS_KEY_ID",
    "AWS_SECRET_ACCESS_KEY",
    "AWS_SESSION_TOKEN",
}

OPENCODE_CAPTURE_TIMEOUT_SECONDS = 30


def run_opencode_command(
    args: list[str],
    *,
    opencode_home: Path,
    launch_cwd: Path,
    timeout: float = OPENCODE_CAPTURE_TIMEOUT_SECONDS,
    env_base: Mapping[str, str] | None = None,
) -> subprocess.CompletedProcess[str]:
    """Run an opencode CLI command with stdout redirected to a file.

    The opencode binary ends every command with a bare process.exit(), which
    discards stdout that has not yet drained. Through a pipe, payloads >64KiB
    arrive truncated at the pipe-buffer boundary (still exit 0) and tiny
    replies can vanish entirely under load. A regular-file stdout drains
    synchronously, so the payload survives. stderr stays piped (always small).
    """
    with tempfile.TemporaryFile(mode="w+", encoding="utf-8") as stdout_file:
        result = subprocess.run(
            ["opencode", *args],
            cwd=launch_cwd,
            text=True,
            stdout=stdout_file,
            stderr=subprocess.PIPE,
            env=opencode_run_env(opencode_home, env_base=env_base),
            timeout=timeout,
        )
        stdout_file.seek(0)
        result.stdout = stdout_file.read()
    return result


def opencode_run_env(
    opencode_home: Path,
    env_base: Mapping[str, str] | None = None,
) -> dict[str, str]:
    source = dict(env_base) if env_base is not None else dict(os.environ)
    allowlist = set(OPENCODE_ENV_ALLOWLIST)
    source_managed = is_managed_host(source) or is_managed_host(os.environ)
    if source_managed:
        target_provider_env = {
            name.strip()
            for name in source.get("QUORUM_TARGET_ENV_KEYS", "").split(",")
            if name.strip()
        }
        allowlist -= OPENCODE_PROVIDER_ENV_NAMES
        allowlist |= target_provider_env & OPENCODE_PROVIDER_ENV_NAMES
    env = {key: value for key, value in source.items() if key in allowlist}
    env.setdefault("PATH", os.defpath)
    env.setdefault("TERM", "xterm-256color")
    env.setdefault("LANG", "C.UTF-8")
    env.update(opencode_env(opencode_home))
    return env


def _realpath(value: str | Path) -> str:
    return os.path.realpath(str(value))


def _session_decisions(
    raw_sessions: Any, launch_cwd: Path
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    if not isinstance(raw_sessions, list):
        raise OpenCodeCaptureError("opencode session list returned non-list JSON")
    target = _realpath(launch_cwd)
    decisions: list[dict[str, Any]] = []
    matches: list[dict[str, Any]] = []
    for index, session in enumerate(raw_sessions):
        if not isinstance(session, dict):
            decisions.append({"index": index, "matched": False, "reason": "non-dict row"})
            continue
        session_row = cast(dict[str, Any], session)
        directory = session_row.get("directory")
        session_id = session_row.get("id")
        if not isinstance(directory, str) or not isinstance(session_id, str):
            decisions.append(
                {
                    "index": index,
                    "id": session_id,
                    "matched": False,
                    "reason": "missing id or directory",
                }
            )
            continue
        directory_realpath = _realpath(directory)
        matched = directory_realpath == target
        decisions.append(
            {
                "index": index,
                "id": session_id,
                "directory": directory,
                "directory_realpath": directory_realpath,
                "launch_cwd_realpath": target,
                "matched": matched,
            }
        )
        if matched:
            matches.append(session_row)
    return decisions, matches


def _list_sessions(
    *,
    opencode_home: Path,
    launch_cwd: Path,
    env_base: Mapping[str, str] | None = None,
) -> list[Any]:
    try:
        result = run_opencode_command(
            ["session", "list", "--format", "json"],
            opencode_home=opencode_home,
            launch_cwd=launch_cwd,
            env_base=env_base,
        )
    except subprocess.TimeoutExpired as e:
        raise OpenCodeCaptureError(
            f"opencode session list timed out after {OPENCODE_CAPTURE_TIMEOUT_SECONDS}s"
        ) from e
    if result.returncode != 0:
        raise OpenCodeCaptureError(
            "opencode session list failed "
            f"(exit {result.returncode}): {result.stderr.strip()[:300]}"
        )
    try:
        sessions = json.loads(result.stdout or "[]")
    except json.JSONDecodeError as e:
        raise OpenCodeCaptureError("opencode session list returned invalid JSON") from e
    if not isinstance(sessions, list):
        raise OpenCodeCaptureError("opencode session list returned non-list JSON")
    return sessions


def snapshot_opencode_sessions(
    *,
    opencode_home: Path,
    launch_cwd: Path,
    env_base: Mapping[str, str] | None = None,
) -> set[str]:
    _decisions, sessions = _session_decisions(
        _list_sessions(opencode_home=opencode_home, launch_cwd=launch_cwd, env_base=env_base),
        launch_cwd,
    )
    return {session["id"] for session in sessions}


def _session_created(session: dict[str, Any]) -> int | None:
    for key in ("created", "time_created"):
        value = session.get(key)
        if isinstance(value, int):
            return value
    return None


def _export_session(
    *,
    session_id: str,
    opencode_home: Path,
    launch_cwd: Path,
    env_base: Mapping[str, str] | None = None,
) -> tuple[dict[str, Any], str, str]:
    try:
        result = run_opencode_command(
            ["export", session_id],
            opencode_home=opencode_home,
            launch_cwd=launch_cwd,
            env_base=env_base,
        )
    except subprocess.TimeoutExpired as e:
        raise OpenCodeCaptureError(
            f"opencode export {session_id} timed out after {OPENCODE_CAPTURE_TIMEOUT_SECONDS}s"
        ) from e
    if result.returncode != 0:
        raise OpenCodeCaptureError(
            f"opencode export {session_id} failed "
            f"(exit {result.returncode}): {result.stderr.strip()[:300]}"
        )
    try:
        exported_json = json.loads(result.stdout)
    except json.JSONDecodeError as e:
        raise OpenCodeCaptureError(
            f"opencode export {session_id} returned invalid JSON "
            f"({len(result.stdout.encode('utf-8', errors='replace'))} bytes; "
            f"head: {result.stdout[:120]!r}; "
            f"stderr: {result.stderr.strip()[:300]})"
        ) from e
    exported_id = exported_json.get("info", {}).get("id")
    if exported_id != session_id:
        raise OpenCodeCaptureError(
            f"opencode export {session_id} returned session id {exported_id!r}"
        )
    return exported_json, result.stdout, result.stderr


def _exported_created(exported_json: dict[str, Any]) -> int | None:
    created = exported_json.get("info", {}).get("time", {}).get("created")
    return created if isinstance(created, int) else None


def export_opencode_sessions(
    *,
    opencode_home: Path,
    export_dir: Path,
    launch_cwd: Path,
    snapshot: set[str],
    env_base: Mapping[str, str] | None = None,
) -> tuple[Path, ...]:
    """Export OpenCode sessions for launch_cwd into export_dir."""
    export_dir.mkdir(parents=True, exist_ok=True)
    raw_sessions = _list_sessions(
        opencode_home=opencode_home,
        launch_cwd=launch_cwd,
        env_base=env_base,
    )
    decisions, sessions = _session_decisions(raw_sessions, launch_cwd)
    new_sessions = [session for session in sessions if session["id"] not in snapshot]
    export_records: list[dict[str, Any]] = []
    for session in new_sessions:
        session_id = session["id"]
        exported_json, stdout, stderr = _export_session(
            session_id=session_id,
            opencode_home=opencode_home,
            launch_cwd=launch_cwd,
            env_base=env_base,
        )
        created = _session_created(session) or _exported_created(exported_json)
        export_records.append(
            {
                "id": session_id,
                "json": exported_json,
                "stdout": stdout,
                "stderr": stderr,
                "created": created,
            }
        )
    if len(export_records) > 1 and any(record["created"] is None for record in export_records):
        raise OpenCodeCaptureError(
            "cannot order multiple new OpenCode sessions without creation times"
        )
    export_records.sort(key=lambda record: (record["created"] or 0, record["id"]))

    exported: list[Path] = []
    manifest: dict[str, Any] = {
        "raw_session_rows": raw_sessions,
        "session_decisions": decisions,
        "snapshot_ids": sorted(snapshot),
        "all_matching_ids": [session["id"] for session in sessions],
        "matched_ids": [session["id"] for session in new_sessions],
        "skipped_existing_ids": [
            session["id"] for session in sessions if session["id"] in snapshot
        ],
        "skipped_nonmatching_ids": [
            decision["id"]
            for decision in decisions
            if decision.get("id") and not decision.get("matched")
        ],
        "exports": [],
    }
    for record in export_records:
        created = record["created"] or 0
        out_path = export_dir / f"{created:016d}-{record['id']}.json"
        out_path.write_text(record["stdout"])
        manifest["exports"].append(
            {
                "id": record["id"],
                "created": created,
                "path": str(out_path),
                "stderr": record["stderr"],
            }
        )
        exported.append(out_path)

    (export_dir / "opencode-session-export-manifest.json").write_text(
        json.dumps(manifest, indent=2) + "\n"
    )
    return tuple(exported)
